network/iptables.go

```diff
@@ -14,152 +14,189 @@ import (
+	// add cgroup path
+	if cgroupPath != "" {
+		rulespec = append(rulespec, "-m", "cgroup", "--path", cgroupPath)
+	}
 
-	rulespec = append([]string{"-m", "cgroup", "--path", cgroupPath}, rulespec...)
+	rulespec = append(rulespec, "-j", chaosChainName)
 
-	return i.ip.AppendUnique(table, chain, rulespec...)
-}
+	// inject output rule
+	if err := i.insert("nat", "OUTPUT", rulespec...); err != nil {
+		return err
+	}
 
-func (i iptables) AddWideFilterRule(chain string, protocol string, port string, jump string) error {
-	if i.dryRun {
-		return nil
+	// inject prerouting rule only if there's no cgroup path filtering
+	// packets going through prerouting chain are not yet associated
+	// to a process so there's no possibility to filter on a cgroup at this stage
+	if cgroupPath == "" {
+		if err := i.insert("nat", "PREROUTING", rulespec...); err != nil {
+			return err
+		}
 	}
 
-	i.log.Infow("creating new iptables rule", "chain name", chain, "protocol", protocol, "port", port, "jump target", jump)
+	return nil
+}
 
-	return i.ip.AppendUnique("nat", chain, "-p", protocol, "--dport", port, "-j", jump)
+// Mark marks the packets in the injector dedicated chain with the given mark
+func (i *iptables) Mark(cgroupPath string, mark string) error {
+	return i.insert("mangle", "OUTPUT", "-m", "cgroup", "--path", cgroupPath, "-j", "MARK", "--set-mark", mark)
 }
 
-func (i iptables) DeleteRule(chain string, protocol string, port string, jump string) error {
+// insert creates a new iptables rule definition, stores it
+// for further cleanup and inserts the rule in the given table and chain
+// at the first position
+func (i *iptables) insert(table string, chain string, rulespec ...string) error {
+	i.log.Infow("injecting iptables rule", "table", table, "chain", chain, "rulespec", rulespec)
+
 	if i.dryRun {
 		return nil
 	}
 
-	// Why do we check if the jump target exists? A command of the form
-	// iptables -t nat -C OUTPUT -p udp --dport 53 -j CHAOS-DNS
-	// will actually error if the jump target does not exist. However, you are unable
-	// to delete a chain if there are rules that jump to it, so if the target does not exist
-	// we can be sure that the rule does not exist.
-	if exists, _ := i.ip.ChainExists("nat", jump); !exists {
-		return nil
+	// create the injector chain if it does not exist yet and is used here
+	if chain == chaosChainName {
+		chainExists, err := i.ip.ChainExists(table, chain)
+		if err != nil {
+			return err
+		}
+
+		if !chainExists {
+			if err := i.ip.NewChain(table, chain); err != nil {
+				return fmt.Errorf("error creating chain %s: %w", chain, err)
+			}
+		}
 	}
 
-	return i.DeleteRuleSpec("nat", chain, "-p", protocol, "--dport", port, "-j", jump)
-}
-
-func (i iptables) DeleteRuleSpec(chain string, rulespec ...string) error {
-	if i.dryRun {
-		return nil
+	// check if the rule already exists before trying to insert it
+	exists, err := i.ip.Exists(table, chain, rulespec...)
+	if err != nil {
+		return err
```

Comment:
shouldn't this be the second thing we do (after `dryrun` check?)

Reply:
If the given chain does not exist yet (which is the case for CHAOS-DNS when first called), the `Exists` function will return an error, this is why we first create the chain if not done yet (note that it's only for the CHAOS-DNS chain, otherwise we'll still error but it's not expected to have an unexisting chain outside of the this one).